fix for properties that have a json object as a value

samuelbirch opened the pull request.

johnnynotsolucky commented:
Thank you @samuelbirch!

I'm looking at this now and the related issue #5. Would you be able to give me a sample of the data which causes the original error, and also a stacktrace? I want to replicate the issue so I can verify your changes.

samuelbirch commented:
I'll look to dig the info out for you but simply i'm added json as a value of a lineItem option

samuelbirch commented:
2019-07-09 20:51:11 [-][-][-][error][fostercommerce\shipstationconnect\controllers\OrdersController::logException] Error processing action export
2019-07-09 20:51:11 [-][-][-][error][yii\base\ErrorException:2] yii\base\ErrorException: DOMDocument::createCDATASection() expects parameter 1 to be string, array given in /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php:500
Stack trace:
#0 /home/forge/project/vendor/craftcms/cms/src/web/ErrorHandler.php(81): yii\base\ErrorHandler->handleError(2, 'DOMDocument::cr...', '/home/forge/pro...', 500)
#1 [internal function]: craft\web\ErrorHandler->handleError(2, 'DOMDocument::cr...', '/home/forge/pro...', 500, Array)
#2 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(500): DOMDocument->createCDATASection(Array)
#3 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(271): fostercommerce\shipstationconnect\services\Xml->addChildWithCDATA(Object(SimpleXMLElement), 'Value', Array)
#4 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(202): fostercommerce\shipstationconnect\services\Xml->options(Object(SimpleXMLElement), Array)
#5 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(138): fostercommerce\shipstationconnect\services\Xml->item(Object(SimpleXMLElement), Object(craft\commerce\models\LineItem))
#6 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(97): fostercommerce\shipstationconnect\services\Xml->items(Object(SimpleXMLElement), Array)
#7 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/services/Xml.php(32): fostercommerce\shipstationconnect\services\Xml->order(Object(SimpleXMLElement), Object(craft\commerce\elements\Order))
#8 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/controllers/OrdersController.php(113): fostercommerce\shipstationconnect\services\Xml->orders(Object(SimpleXMLElement), Array)
#9 /home/forge/project/vendor/fostercommerce/shipstationconnect/src/controllers/OrdersController.php(39): fostercommerce\shipstationconnect\controllers\OrdersController->getOrders()
#10 [internal function]: fostercommerce\shipstationconnect\controllers\OrdersController->actionProcess()
#11 /home/forge/project/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#12 /home/forge/project/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#13 /home/forge/project/vendor/craftcms/cms/src/web/Controller.php(109): yii\base\Controller->runAction('process', Array)
#14 /home/forge/project/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('process', Array)
#15 /home/forge/project/vendor/craftcms/cms/src/web/Application.php(297): yii\base\Module->runAction('shipstationconn...', Array)
#16 /home/forge/project/vendor/craftcms/cms/src/web/Application.php(565): craft\web\Application->runAction('shipstationconn...', Array)
#17 /home/forge/project/vendor/craftcms/cms/src/web/Application.php(281): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#18 /home/forge/project/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#19 /home/forge/project/public/index.php(29): yii\base\Application->run()
#20 {main}

samuelbirch commented:
with further testing it seems that the json was hitting the string type character limit, so i've changed it to return an empty string. For my needs i do not need to return the value as it contains extra information coming from a 3rd system that we use to modify the order details.

I've also changed the order query to use dateUpdated instead of dateOrdered when date params are passed in. In my case the date when an order is updated to the status of "paid" is very different to dateOrdered and therefore does not get picked up with the current order query.

johnnynotsolucky commented on src/services/Xml.php:
```diff
@@ -496,7 +496,10 @@ protected function addChildWithCDATA(&$xml, $name, $value)
         $new_child = $xml->addChild($name);
         if ($new_child !== null) {
-            $node = dom_import_simplexml($new_child);
+			$node = dom_import_simplexml($new_child);
```
You should be able to revert this, and instead on line [201-203](https://github.com/FosterCommerce/shipstation-connect/pull/6/files#diff-baff5883d2219e1e7dfe686e4724d017R202), that if statement and its body can be removed. I think a lot of the issue you experienced is to do with the snapshot JSON plus the JSON you've added to order items.

johnnynotsolucky commented on src/controllers/OrdersController.php:
```diff
@@ -99,7 +99,7 @@ protected function getOrders()
 
         if ($start_date && $end_date) {
-            $query->dateOrdered(array('and', '> '.$start_date, '< '.$end_date));
+            $query->dateUpdated(array('and', '> '.$start_date, '< '.$end_date));
```
I'd prefer to keep this as `dateOrdered`. The `dateUpdated` field can is updated for any change to the order and I feel it would be unreliable for a large number of use cases.

samuelbirch replied:
Thats fine, i'll just use this forked version until your new version is ready. If you could you add in the ability to customise/change the order query from another plugin that would be great.

johnnynotsolucky commented:
@samuelbirch you were correct with regards to updating to `dateUpdated`. The ShipStation docs explicitly mention filtering your orders on the modified date. I've made change you made, and a few others, but I need to test a bit more thoroughly. I'll probably update early next week.

samuelbirch commented:
ah thank you :)
